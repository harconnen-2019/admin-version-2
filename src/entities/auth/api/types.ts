import { z } from 'zod';

import { placeType } from '@/entities/places';
import { schemaValidate } from '@/shared/api';

export const schemaSession = z
  .object({
    session: z
      .object({
        user: z.object({ id: z.number(), username: z.string() }),
        place: placeType.schemaPlace,
      })
      .optional(),
  })
  .merge(schemaValidate);

export interface IRequestLogin {
  username: string;
  password: string;
  saveSession?: boolean;
}
