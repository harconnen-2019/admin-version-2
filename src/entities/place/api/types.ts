import { schemaDate, schemaValidate } from '@/shared/api';
import { z } from 'zod';

/**
 * Схема элемента витрины
 * Поле type является элементом другой схемы
 * Добавляем зарезервированные поля с датами
 */
export const schemaPlace = z
  .object({
    id: z.number(),
    name: z.string(),
    domain: z.string(),
    template: z.string(),
    // TODO вставить схему type
    type: z.any(),
    favicon: z.string().nullable(),
    og_img: z.string().nullable(),
    logo_light: z.string().nullable(),
    logo_dark: z.string().nullable(),
    color_scheme: z.string().nullable(),
    counter_head: z.string(),
    counter_body: z.string(),
    thankyou_type: z.enum(['', 'pop', 'page', 'off']),
  })
  .merge(schemaDate);

export type IPlace = z.infer<typeof schemaPlace>;

/**
 * Схема получения API для вывода одной витрины с подключенной валидацией
 * Также применяется при получении API при добавлении и редактировании
 */
export const schemaGetPlace = z
  .object({
    places_item: schemaPlace.optional(),
  })
  .merge(schemaValidate);

/**
 * Схема получения API для вывода списка витрин с подключенной валидацией
 */
export const schemaListPlace = z
  .object({
    places_item_list: z.array(schemaPlace).optional(),
  })
  .merge(schemaValidate);

/**
 * Схема отправки API для добавления новой витрины
 * Удаляем поля, которые не применяются
 * Переопределяем тип некоторых полей
 */
const schemaRequestPostPlace = schemaPlace
  .omit({
    created: true,
    modified: true,
    id: true,
  })
  .extend({
    favicon: z.instanceof(File).optional(),
    og_img: z.instanceof(File).optional(),
    logo_light: z.instanceof(File).optional(),
    logo_dark: z.instanceof(File).optional(),
    type: z.number(),
  });
export type IRequestPostPlace = z.infer<typeof schemaRequestPostPlace>;

/**
 * Схема отправки API для редактирования витрины
 * Это копия схемы добавления + добавлен ID редактируемой витрины
 */
const schemaRequestPutPlace = schemaRequestPostPlace.extend({
  id: z.number(),
});
export type IRequestPutPlace = z.infer<typeof schemaRequestPutPlace>;
