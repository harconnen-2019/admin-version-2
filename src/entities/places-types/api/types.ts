import { schemaDate, schemaValidate } from '@/shared/api';
import { z } from 'zod';

/**
 * Схема элемента типа витрины
 * Добавляем зарезервированные поля с датами
 */
export const schemaPlacesType = z
  .object({
    id: z.number(),
    name: z.string(),
  })
  .merge(schemaDate);

export type IPlacesType = z.infer<typeof schemaPlacesType>;

/**
 * Схема получения API для вывода одного тип витрины с подключенной валидацией
 * Также применяется при получении API при добавлении и редактировании
 */
export const schemaGetPlaceType = z
  .object({
    places_type: schemaPlacesType.optional(),
  })
  .merge(schemaValidate);

/**
 * Схема получения API для вывода списка типов витрин с подключенной валидацией
 */
export const schemaListPlaceType = z
  .object({
    places_type_list: z.array(schemaPlacesType).optional(),
  })
  .merge(schemaValidate);
