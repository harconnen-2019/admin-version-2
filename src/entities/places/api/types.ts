import { z } from 'zod';

import { typeFromPlacesType } from '@/entities/types-from-places';
import { schemaDate, schemaValidate } from '@/shared/api';

/**
 * Схема элемента витрины
 * Поле 'type' является элементом другой схемы
 * Добавляем зарезервированные поля с датами
 */
export const schemaPlace = z
  .object({
    id: z.number(),
    name: z.string(),
    domain: z.string(),
    template: z.string(),
    type: typeFromPlacesType.schemaTypeFromPlaces,
    favicon: z.string().nullable(),
    og_img: z.string().nullable(),
    logo_light: z.string().nullable(),
    logo_dark: z.string().nullable(),
    color_scheme: z.string().nullable(),
    counter_head: z.string(),
    counter_body: z.string(),
    thankyou_type: z.enum(['', 'pop', 'page', 'off']),
  })
  .merge(schemaDate);

/**
 * Схема получения данных api для вывода одной витрины с подключенной валидацией
 * Также применяется при получения данных api, при добавлении и редактировании
 */
export const schemaGetPlace = z
  .object({
    places_item: schemaPlace.optional(),
  })
  .merge(schemaValidate);

/**
 * Схема получения данных api для вывода списка витрин с подключенной валидацией
 */
export const schemaListPlace = z
  .object({
    places_item_list: z.array(schemaPlace).optional(),
  })
  .merge(schemaValidate);

/**
 * Схема отправки данных api для добавления новой витрины
 * Удаляем поля, которые не применяются
 * Переопределяем тип некоторых полей
 */
const schemaRequestPostPlace = schemaPlace
  .omit({
    created: true,
    modified: true,
    id: true,
  })
  .extend({
    favicon: z.union([z.string(), z.instanceof(File)]).optional(),
    og_img: z.union([z.string(), z.instanceof(File)]).optional(),
    logo_light: z.union([z.string(), z.instanceof(File)]).optional(),
    logo_dark: z.union([z.string(), z.instanceof(File)]).optional(),
    type: z.number(),
  });

/**
 * Схема отправки api для редактирования витрины
 * Это копия схемы добавления + добавлен id редактируемой витрины
 */
const schemaRequestPutPlace = schemaRequestPostPlace.extend({
  id: z.number(),
});

/**
 * Экспортируемые типы
 */

export type IPlace = z.infer<typeof schemaPlace>;
export type IRequestPostPlace = z.infer<typeof schemaRequestPostPlace>;
export type IRequestPutPlace = z.infer<typeof schemaRequestPutPlace>;
