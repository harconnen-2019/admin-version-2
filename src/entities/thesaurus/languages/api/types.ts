import { schemaDate, schemaValidate } from '@/shared/api';
import { z } from 'zod';

/**
 * Схема элемента Языка
 * Добавляем зарезервированные поля с датами
 */
export const schemaLanguage = z
  .object({
    id: z.number(),
    name: z.string(),
    slug: z.string(),
  })
  .merge(schemaDate);

/**
 * Схема получения API для вывода одного языка с подключенной валидацией
 * Также применяется при получении API при добавлении и редактировании
 */
export const schemaGetLanguage = z
  .object({
    thesaurus_language: schemaLanguage.optional(),
  })
  .merge(schemaValidate);

/**
 * Схема получения API для вывода списка языков с подключенной валидацией
 */
export const schemaListLanguage = z
  .object({
    thesaurus_language_list: z.array(schemaLanguage).optional(),
  })
  .merge(schemaValidate);

/**
 * Схема отправки данных api для добавления нового языка
 * Удаляем поля, которые не применяются
 * Переопределяем тип некоторых полей
 */
const schemaRequestPostLanguage = schemaLanguage.omit({
  created: true,
  modified: true,
  id: true,
});

/**
 * Схема отправки api для редактирования языка
 * Это копия схемы добавления + добавлен id редактируемого языка
 */
const schemaRequestPutLanguage = schemaRequestPostLanguage.extend({
  id: z.number(),
});

/**
 * Экспортируемые типы
 */

export type ILanguage = z.infer<typeof schemaLanguage>;
export type IRequestPostLanguage = z.infer<typeof schemaRequestPostLanguage>;
export type IRequestPutLanguage = z.infer<typeof schemaRequestPutLanguage>;
