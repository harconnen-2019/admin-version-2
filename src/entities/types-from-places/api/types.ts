import { z } from 'zod';

import { schemaDate, schemaValidate } from '@/shared/api';

/**
 * Схема элемента типа витрины
 * Добавляем зарезервированные поля с датами
 */
export const schemaTypeFromPlaces = z
  .object({
    id: z.number(),
    name: z.string(),
  })
  .merge(schemaDate);

export type ITypeFromPlaces = z.infer<typeof schemaTypeFromPlaces>;

/**
 * Схема получения API для вывода одного тип витрины с подключенной валидацией
 * Также применяется при получении API при добавлении и редактировании
 */
export const schemaGetTypeFromPlaces = z
  .object({
    places_type: schemaTypeFromPlaces.optional(),
  })
  .merge(schemaValidate);

/**
 * Схема получения API для вывода списка типов витрин с подключенной валидацией
 */
export const schemaListTypeFromPlaces = z
  .object({
    places_type_list: z.array(schemaTypeFromPlaces).optional(),
  })
  .merge(schemaValidate);
