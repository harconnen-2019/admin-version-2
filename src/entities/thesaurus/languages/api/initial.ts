import { defaultValidate } from '@/shared/lib';

import { ILanguage, IRequestPostLanguage, IRequestPutLanguage } from './types';

/**
 * Методы для инициализации форм создания и редактирования
 */

const initial: IRequestPostLanguage | IRequestPutLanguage = {
  name: '',
  slug: '',
};

// какие поля будут проверятся на заполнение (обязательные)
export const languageFormValidate = {
  name: defaultValidate,
  slug: defaultValidate,
};

// создание
export const languageInitialPost = (): IRequestPostLanguage => {
  return initial;
};

// редактирование, id проверяется на странице
export const languageInitialPut = (id: string | number): IRequestPutLanguage => {
  const result = { ...(initial as IRequestPutLanguage) };
  result.id = Number(id);

  return result;
};

// присвоение данных при загрузки с сервера
export const languageLoadFromApi = (api: ILanguage) => {
  const result = { ...(initial as IRequestPutLanguage) };
  for (const key in result) {
    result[key as keyof IRequestPutLanguage] = api[key as keyof ILanguage] as never;
  }
  return result;
};
