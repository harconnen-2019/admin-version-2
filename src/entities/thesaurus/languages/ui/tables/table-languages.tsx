import { Badge, Table } from '@mantine/core';

import { PATH } from '@/pages/path';
import { getDate } from '@/shared/lib';
import { ButtonDell, LinkAnchor, TableData } from '@/shared/ui';

import { ILanguage } from '../../api/types';

interface IProperties {
  status: string;
  error: Error | null | undefined;
  data: ILanguage[];
  remove: (id: number, name: string) => void;
}

/**
 * Таблица с данными языков
 * @param root0 пропсы
 * @param root0.data список данных
 * @param root0.error ошибка
 * @param root0.status статус загрузки
 * @param root0.remove метод удаления
 * @returns JSX Element
 */
export function TableLanguages({ data, error, status, remove }: Readonly<IProperties>) {
  /**
   * Формируем строки таблицы
   */
  const rows = data.map((element) => (
    <Table.Tr key={element.id}>
      <Table.Td>
        <Badge variant="default">{element.slug}</Badge>
      </Table.Td>
      <Table.Td>
        <LinkAnchor to={PATH.thesaurus.languages.edit(element.id)}>{element.name}</LinkAnchor>
      </Table.Td>
      <Table.Td>{getDate(element.created)}</Table.Td>
      <Table.Td>{getDate(element.modified)}</Table.Td>
      <Table.Td align="right">
        <ButtonDell callback={() => remove(element.id, element.name)} />
      </Table.Td>
    </Table.Tr>
  ));

  return (
    <TableData
      isLoading={status === 'pending'}
      error={error}
      tableHead={[
        { id: 1, name: 'Slug', w: 50 },
        { id: 2, name: 'Язык', w: 'auto' },
        { id: 3, name: 'Создан', w: 170 },
        { id: 4, name: 'Изменен', w: 170 },
        { id: 5, name: ' ', w: 50 },
      ]}
      empty={(status === 'success') === (data.length === 0)}
    >
      {status === 'success' && data.length > 0 && rows}
    </TableData>
  );
}
