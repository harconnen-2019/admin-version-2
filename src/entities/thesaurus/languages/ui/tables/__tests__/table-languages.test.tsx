import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render, screen } from '@test-utils';
import { BrowserRouter } from 'react-router-dom';
import { describe, it } from 'vitest';

import { ILanguage } from '../../../api/types';
import { TableLanguages } from '../table-languages';

const data: ILanguage[] = [
  {
    id: 1,
    created: new Date('1946-10-04T18:12:23.671Z'),
    modified: new Date('1948-06-03T12:04:33.301Z'),
    name: 'Русский',
    slug: 'ru',
  },
  {
    id: 2,
    created: new Date('1948-10-04T18:12:23.671Z'),
    modified: new Date('1948-06-03T12:04:33.301Z'),
    name: 'English',
    slug: 'en',
  },
];

const queryClient = new QueryClient();

describe('Таблица со списком языков', () => {
  const renderComponent = () =>
    render(
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          <TableLanguages data={data} error={undefined} remove={() => {}} status="success" />
        </BrowserRouter>
      </QueryClientProvider>,
    );

  it('Создаем "снап-шот" данных', () => {
    expect(data).toMatchSnapshot();
  });

  it('Данные появятся только если статус "success', () => {
    render(
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          <TableLanguages data={data} error={undefined} remove={() => {}} status="pending" />
        </BrowserRouter>
      </QueryClientProvider>,
    );
    const result = screen.queryByText(/clothing/i);

    expect(result).not.toBeInTheDocument();
  });

  it('Дата переводится в удобный формат', () => {
    renderComponent();
    const parseDate = screen.getByText(/4 октября 46 г/i);

    expect(parseDate).toBeInTheDocument();
  });

  it('Адаптивная ширина у колонки с названием', () => {
    renderComponent();
    const cellHeadName = screen.getByRole('columnheader', {
      name: /язык/i,
    });

    expect(cellHeadName).toHaveAttribute('style', 'width: auto;');
  });
});
