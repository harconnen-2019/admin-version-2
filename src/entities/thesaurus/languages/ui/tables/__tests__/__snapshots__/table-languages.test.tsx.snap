// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Таблица со списком языков > Создаем "снап-шот" данных 1`] = `
[
  {
    "created": 1946-10-04T18:12:23.671Z,
    "id": 1,
    "modified": 1948-06-03T12:04:33.301Z,
    "name": "Русский",
    "slug": "ru",
  },
  {
    "created": 1948-10-04T18:12:23.671Z,
    "id": 2,
    "modified": 1948-06-03T12:04:33.301Z,
    "name": "English",
    "slug": "en",
  },
]
`;
